File: src/test/java/uk/q3c/krail/core/navigate/DefaultLoginNavigationRuleTest.java
/*
 * Copyright (c) 2015. David Sowerby
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with
 * the License. You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
 * an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
 * specific language governing permissions and limitations under the License.
 */

package uk.q3c.krail.core.navigate;

import com.mycila.testing.junit.MycilaJunitRunner;
import com.mycila.testing.plugin.guice.GuiceContext;
import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.mockito.Mock;
import uk.q3c.krail.core.navigate.sitemap.StandardPageKey;
import uk.q3c.krail.core.navigate.sitemap.UserSitemap;
import uk.q3c.krail.core.navigate.sitemap.UserSitemapNode;
import uk.q3c.krail.core.user.UserStatusChangeSource;

import java.util.Optional;

import static org.assertj.core.api.Assertions.assertThat;
import static org.mockito.Mockito.any;
import static org.mockito.Mockito.when;

@RunWith(MycilaJunitRunner.class)
@GuiceContext({})
public class DefaultLoginNavigationRuleTest {

    @Mock
    Navigator navigator;

    @Mock
    UserStatusChangeSource source;

    DefaultLoginNavigationRule rule;
    @Mock
    UserSitemapNode aPageNode;
    @Mock
    UserSitemapNode loginNode;
    @Mock
    NavigationState previousNavigationState;
    @Mock
    NavigationState homeNavigationState;
    @Mock
    private UserSitemapNode logoutNode;
    @Mock
    private UserSitemapNode previousNode;
    @Mock
    private URIFragmentHandler uriHandler;
    @Mock
    private UserSitemap userSitemap;

    @Before
    public void setup() {
        rule = new DefaultLoginNavigationRule(userSitemap, uriHandler);
        when(userSitemap.standardPageNode(StandardPageKey.Log_In)).thenReturn(loginNode);
        when(userSitemap.standardPageNode(StandardPageKey.Log_Out)).thenReturn(logoutNode);
    }

    @Test
    public void not_on_login_page() {
        //given
        when(navigator.getCurrentNode()).thenReturn(aPageNode);
        //when
        Optional<NavigationState> expected = rule.changedNavigationState(navigator, source);
        //then
        assertThat(expected.isPresent()).isFalse();
    }

    @Test
    public void on_login_page_no_previous() {
        //given
        when(navigator.getCurrentNode()).thenReturn(loginNode);
        when(navigator.getPreviousNode()).thenReturn(null);
        when(uriHandler.navigationState(any(String.class))).thenReturn(homeNavigationState);
        //when
        Optional<NavigationState> expected = rule.changedNavigationState(navigator, source);
        //then
        assertThat(expected.isPresent()).isTrue();
        assertThat(expected.get()).isEqualTo(homeNavigationState);
    }

    @Test
    public void login_page_has_previous() {
        //given
        when(navigator.getCurrentNode()).thenReturn(loginNode);
        when(navigator.getPreviousNode()).thenReturn(previousNode);
        when(navigator.getPreviousNavigationState()).thenReturn(previousNavigationState);
        //when
        Optional<NavigationState> expected = rule.changedNavigationState(navigator, source);
        //then
        assertThat(expected.isPresent()).isTrue();
        assertThat(expected.get()).isEqualTo(previousNavigationState);
    }

    @Test
    public void on_login_page_previous_was_logout() {
        //given
        when(navigator.getCurrentNode()).thenReturn(loginNode);
        when(navigator.getPreviousNode()).thenReturn(logoutNode);
        when(uriHandler.navigationState(any(String.class))).thenReturn(homeNavigationState);
        //when
        Optional<NavigationState> expected = rule.changedNavigationState(navigator, source);
        //then
        assertThat(expected.isPresent()).isTrue();
        assertThat(expected.get()).isEqualTo(homeNavigationState);
    }
}

File: src/test/kotlin/uk/q3c/krail/core/navigate/DefaultLoginNavigationRuleTest.kt
package uk.q3c.krail.core.navigate

import io.mockk.every
import io.mockk.mockk
import org.amshove.kluent.shouldBe
import org.amshove.kluent.shouldBeFalse
import org.amshove.kluent.shouldBeTrue
import org.jetbrains.spek.api.Spek
import org.jetbrains.spek.api.dsl.given
import org.jetbrains.spek.api.dsl.it
import org.jetbrains.spek.api.dsl.on
import uk.q3c.krail.core.navigate.sitemap.StandardPageKey
import uk.q3c.krail.core.navigate.sitemap.UserSitemap
import uk.q3c.krail.core.navigate.sitemap.UserSitemapNode
import uk.q3c.krail.core.user.status.UserStatusChangeSource

/**
 * Created by David Sowerby on 11 Jun 2018
 */
object DefaultLoginNavigationRuleTest : Spek({

    given("a Login navigation rule") {
        lateinit var navigator: Navigator
        lateinit var source: UserStatusChangeSource
        lateinit var rule: DefaultLoginNavigationRule
        lateinit var loginNode: UserSitemapNode
        lateinit var previousNavigationState: NavigationState
        lateinit var homeNavigationState: NavigationState
        lateinit var currentNavigationState: NavigationState
        lateinit var logoutNode: UserSitemapNode
        lateinit var previousNode: UserSitemapNode
        lateinit var uriHandler: URIFragmentHandler
        lateinit var userSitemap: UserSitemap

        beforeEachTest {
            navigator = mockk(relaxed = true)
            source = mockk(relaxed = true)
            rule = mockk(relaxed = true)
            loginNode = mockk(relaxed = true)
            previousNavigationState = mockk(relaxed = true)
            homeNavigationState = mockk(relaxed = true)
            currentNavigationState = mockk(relaxed = true)
            logoutNode = mockk(relaxed = true)
            previousNode = mockk(relaxed = true)
            uriHandler = mockk(relaxed = true)
            userSitemap = mockk(relaxed = true)

            rule = DefaultLoginNavigationRule(userSitemap, uriHandler)

            every { userSitemap.standardPageNode(StandardPageKey.Log_In) } returns loginNode
            every { userSitemap.standardPageNode(StandardPageKey.Log_Out) } returns logoutNode
            every { navigator.currentNavigationState } returns currentNavigationState
        }

        on("not on login page") {
            every { userSitemap.isLoginUri(navigator.currentNavigationState) } returns false
            val expected = rule.changedNavigationState(navigator, source)

            it("evaluates to false") {
                expected.isPresent.shouldBeFalse()
            }
        }



        on(" on_login_page_no_previous") {
            //given
            every { userSitemap.isLoginUri(navigator.currentNavigationState) } returns (true)
            every { navigator.previousNode } returns (null)
            every { uriHandler.navigationState(any<String>()) } returns (homeNavigationState)
            //when
            val expected = rule.changedNavigationState(navigator, source)
            //then
            expected.isPresent.shouldBeTrue()
            expected.get().shouldBe(homeNavigationState)
        }


        on(" login_page_has_previous") {
            //given

            every { userSitemap.isLoginUri(navigator.currentNavigationState) } returns (true)
            every { navigator.previousNode } returns (previousNode)
            every { navigator.previousNavigationState } returns (previousNavigationState)
            //when
            val expected = rule.changedNavigationState(navigator, source)
            //then
            expected.isPresent.shouldBeTrue()
            expected.get().shouldBe(previousNavigationState)
        }


        on(" on_login_page_previous_was_logout") {
            //given
            every { userSitemap.isLoginUri(navigator.currentNavigationState) } returns (true)
            every { navigator.previousNode } returns (logoutNode)
            every { uriHandler.navigationState(any<String>()) } returns (homeNavigationState)
            //when
            val expected = rule.changedNavigationState(navigator, source)
            //then
            expected.isPresent.shouldBeTrue()
            expected.get().shouldBe(homeNavigationState)
        }
    }
}) 


    
